video/out/d3d11/context.c

```diff
@@ -141,15 +148,127 @@ static bool d3d11_submit_frame(struct ra_swapchain *sw,
+    hr = IDXGISwapChain_GetFrameStatistics(p->swapchain, &stats);
+    if (hr == DXGI_ERROR_FRAME_STATISTICS_DISJOINT) {
+        p->last_sync_refresh_count = 0;
+        p->last_sync_qpc_time = 0;
+        p->vsync_duration_qpc = 0;
+    }
+    if (FAILED(hr))
+        return;
+
+    // Detecting skipped vsyncs is possible but not supported yet
+    info->skipped_vsyncs = 0;
+
+    // Get the number of physical vsyncs that have passed since the last call.
+    // Check for 0 here, since sometimes GetFrameStatistics returns S_OK but
+    // with 0s in some (all?) members of DXGI_FRAME_STATISTICS.
+    unsigned src_passed = 0;
+    if (stats.SyncRefreshCount && p->last_sync_refresh_count)
+        src_passed = stats.SyncRefreshCount - p->last_sync_refresh_count;
+    p->last_sync_refresh_count = stats.SyncRefreshCount;
+
+    // Get the elapsed time passed between the above vsyncs
+    unsigned sqt_passed = 0;
+    if (stats.SyncQPCTime.QuadPart && p->last_sync_qpc_time)
+        sqt_passed = stats.SyncQPCTime.QuadPart - p->last_sync_qpc_time;
+    p->last_sync_qpc_time = stats.SyncQPCTime.QuadPart;
+
+    // If any vsyncs have passed, estimate the physical frame rate
+    if (src_passed && sqt_passed)
+        p->vsync_duration_qpc = sqt_passed / src_passed;
+    if (p->vsync_duration_qpc)
+        info->vsync_duration = qpc_to_us(sw, p->vsync_duration_qpc);
+
+    // If the physical frame rate is known and the other members of
+    // DXGI_FRAME_STATISTICS are non-0, estimate the timing of the next frame
+    if (p->vsync_duration_qpc && stats.PresentCount &&
+        stats.PresentRefreshCount && stats.SyncRefreshCount &&
+        stats.SyncQPCTime.QuadPart)
+    {
+        // PresentRefreshCount and SyncRefreshCount might refer to different
+        // frames (this can definitely occur in bitblt-mode.) Assuming mpv
+        // presents on every frame, guess the present count that relates to
+        // SyncRefreshCount.
+        unsigned expected_sync_pc = stats.PresentCount +
+            (stats.SyncRefreshCount - stats.PresentRefreshCount);
+
+        // Now guess the timestamp of the last submitted frame based on the
+        // timestamp of the frame at SyncRefreshCount and the frame rate
+        int64_t last_queue_display_time_qpc = stats.SyncQPCTime.QuadPart +
+            (submit_count - expected_sync_pc) * p->vsync_duration_qpc;
+
+        // Only set the estimated display time if it's after the last submission
+        // time. It could be before if mpv skips a lot of frames.
+        if (last_queue_display_time_qpc >= p->last_submit_qpc) {
+            info->last_queue_display_time = mp_time_us() +
+                (qpc_to_us(sw, last_queue_display_time_qpc) - qpc_us_now(sw));
+        }
+    }
+}
+
 static int d3d11_control(struct ra_ctx *ctx, int *events, int request, void *arg)
 {
+    struct priv *p = ctx->priv;
+
     int ret = vo_w32_control(ctx->vo, events, request, arg);
+    if (request == VOCTRL_RESUME) {
```

Comment:
I'd rather see less use of this than more. It's not even a 100% guarantee that video frames come again.

Reply:
Agreed, though I think there needs to be some way for the context to know when the VO intends to start rendering on every vsync. Otherwise I don't think it can tell the difference between a brief pause and a bunch of dropped frames. context_drm_egl.c seems to use VOCTRL_RESUME for something similar.

Reply:
Come to think of it, I could probably remove this without too much of a change in behaviour. The VOCTRL_RESUME arrives one frame too late anyway, which was why I had to add the ``last_queue_display_time_qpc >= p->last_submit_qpc`` sanity check to prevent get_vsync from reporting a ``last_queue_display_time`` before the frame was submitted.